### ISSUE TITLE
Status of Parent task not depending on progress vs.total

### ISSUE THREAD

**Issue by Skrattoune** [boxine/django-huey-monitor#91]
Hi,
My tasks are organised as a set of sub-tasks
![image](https://user-images.githubusercontent.com/56255427/159537483-2bc3faa0-b0ac-4908-892b-a2651765d770.png)
completion of each sub-task is adding 1 to the progress of the parent task

however if you look at the first snapshot that shows only the columns visible without scrolling, one could think that all activities are complete and that it would be ok to switch off the server.

But if you scroll right to see the progression data, the situation is very different:
![image](https://user-images.githubusercontent.com/56255427/159538192-bf242229-ca10-4ada-9600-43316a8b0888.png)

At the moment, the status of the Parent task is not related at all to the task progression (`TaskModel.progress_info[0]` vs. `TaskModel.total`)

In my use case, that would be necessary to take task progression into account when a total is specified.

How could I ensure that the "signal_complete" is not sent before each sub-task is complete?
and that it is sent once the last sub-task has been completed?

**Comment by jedie** [comment]
> How could I ensure that the "signal_complete" is not sent before each sub-task is complete?
and that it is sent once the last sub-task has been completed?

That's on the Huey side: If your main task just fires the sub tasks, then it's completed and Huey send this signal. Don't know if there is something in Huey to recognized this.

Think you have to "wait" in your main task that all sub task are finished... But if you wait, than maybe to ran into a deadlook, if some of the sub tasks has a problem?!?

Quick idea: Maybe the main task can "wait" until all sub tasks signals are one of these: https://github.com/boxine/django-huey-monitor/blob/0373ce9466fad12f3d5164443e45d3869d845dee/huey_monitor/constants.py#L9

**Comment by Skrattoune** [comment]
Thanks Jens,

Exploring further the code for v0.5.0, I had another idea that will require less access to database.
(by the way : great simplification ! congrats)

I'm testing this idea locally since yesterday.
I'll push it between today and week-end.

I'll propose you as well some things I have been using and could prove useful to other users

**Comment by Skrattoune** [comment]
Now you have the different PRs
don't hesitate to adapt to your own style
or to ask clarifications if some things are not self-explaining

I saw also that one of the automatic test didn't succeed,  but I have no clue why.
I'll need your input there.
Have fun !

**Comment by Skrattoune** [comment]
I just realized this morning that in the admin, the parent_task with the sub-task which was latest updated only appeared in 4th position. 
This is not very logic, and it makes it difficult to quickly assess where we are in the execution when a number of tasks have been launched.

 I therefore propose to update `update_dt` of the parent task when a signal is received from a sub-task

**Comment by jedie** [comment]
>  I therefore propose to update update_dt of the parent task when a signal is received from a sub-task
Sounds a good idea!

**Comment by Skrattoune** [comment]
- [ ] we need also to check that the duration of the parent_task is updated according to the last update_dt (which is not the case on the version of huey-monitor which is currently running on my prod system)

I have a currently running task which runs for 10-15h and has a duration of 30.5 seconds :-)

**Comment by jedie** [comment]
> I saw you merged than reverted.
> It didn't work for you ?

Yes, see: https://github.com/boxine/django-huey-monitor/pull/96#issuecomment-1082736005


Think it's all a little complicated.
First of all, we need to distinguish two things:

1. The status of a Task/Subtask (e.g.: "executing", "complete" etc.) -> The current problem, here, isn't it?
2. The progress values and total/percentage/throuput etc. -> For this we have `cumulate2parents=True/False`

Then we have two kind of "Task/Subtask" usage:

1. The maintask divided the work into chunks and start a subtask for every chunk -> `cumulate2parents=True`
2. There are different jobs to do: A maintask starts "independ" subtask. Every subtask has own/different total count -> `cumulate2parents=False`

The first one, is easy and the current supported implementation: All subtask progress are cumulate to the main task. The test project contains a example for this here: 

https://github.com/boxine/django-huey-monitor/blob/fbb974118b5edf53302d54aa6927f459d99bc962/huey_monitor_tests/test_app/tasks.py#L54-L70

For the second (your) case: We didn't have a example in test project. We should add a example to test everything.

About the main-/subtask status: This we have to implement some logic here for all usecases to display the "effective status" of the main task, e.g.:

- If one of the sub task is "executing" -> main task == "executing"
- If all sub tasks are "complete" -> main task "complete"
- If all sub task are "complete" and at least one sub task is "failed" -> main task "failed"

Think this can be done independent in a separated PR.

**Comment by Skrattoune** [comment]
Hi Jedie,

It took me some time to go through. I must say that I don't fully follow your manipulations with PRs, reversion & all.
I tried to cut the things I had implemented on my system in 3 PRs which could be self-difficient.

I did integrate the test project as part of the testing files in PR #97. 
I tried to mimick as much as possible the way you are doing it.


I agree with your analysis, and the last part (main-/subtask status), I'm less confortable with the implementation

**Comment by Skrattoune** [comment]
could you give me a feedback on this?

**Comment by Skrattoune** [comment]
> About the main-/subtask status: This we have to implement some logic here for all usecases to display the "effective status" of the main task, e.g.:
> 
> * If one of the sub task is "executing" -> main task == "executing"
> * If all sub tasks are "complete" -> main task "complete"
> * If all sub task are "complete" and at least one sub task is "failed" -> main task "failed"
> 
> Think this can be done independent in a separated PR.

Hi @jedie,

I think I found an "easy" solution for this:
Signals are associated to task_models using `tasks.store_signals`
I propose to add to `tasks.store_signals` the following lines :

```
        if task_model_instance.parent_task:
            update_task_instance(
                instance=task_model_instance.parent_task,
                last_signal=last_signal,
            )
```
That way, the last_signal will be associated to the parent task in addition to the sub-task
* If all sub tasks are "complete" -> main task "complete" (because the last signal will be "complete")
* If one of the sub task is sending "executing" signal -> main task == "executing" 
* if the last signal from a sub-task is sending "complete" => in most cases, another sub-task is launched, then status wiull turn again to "executing". If not, status is "complete" ... which anyway is the case nowadays
* If one of the sub task is sending "error" signal -> main task == "error" 
with this solution, status "error" only stays if no new signal is associated to the parent task

We could easily prevent this "error" status to be overwritten, for instance by checking, at the begining of `tasks.update_task_instance` 
```
if instance.state.signal_name = 'error':
    return
```
the main drawback for me would be that we are loosing the information that there is still a task running
... and as a consequence it is unwise to kill the huey server, for instance for fixing a bug

What do you think ?
